### PR TITLE
[4.7.5] Implement reaction on ap_metric_interval

### PR DESCRIPTION
**Pull request by VladyslavTupikin**
agent: backhaul_manager: Implement reaction on ap_metric_interval

Certification test 4.7.5 sets ap_metric_interval as part of AP Policy Config message and the agent should send AP Metrics Response every ap_metric_interval seconds until an agent will get another AP Policy Config message with ap_metric_interval equal to 0.

Update the after_select method by adding part of the code for processing ap_metric_interval.

### CODE REVIEW

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 1]
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
         }
     }
 
+    // Send ap metrics query every reporting_interval_s seconds
+    static std::chrono::steady_clock::time_point ap_metric_interval =
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: There was a placeholder for this code in backhaul_fsm_main(), inside case EState::OPERATIONAL

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
...
     // Send topology discovery every 60 seconds according to IEEE_Std_1905.1-2013 specification
     static std::chrono::steady_clock::time_point discovery_timestamp =
         std::chrono::steady_clock::now();
-    auto now     = std::chrono::steady_clock::now();
-    auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - discovery_timestamp);
+    now     = std::chrono::steady_clock::now();
+    elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - discovery_timestamp);
     if ((m_eFSMState == EState::CONNECTED) ||
-        (m_eFSMState == EState::OPERATIONAL && elapsed.count() == 60)) {
+        (m_eFSMState == EState::OPERATIONAL && elapsed.count() >= 60)) {
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This fix should be in a separate commit, even in a separate PR

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 3]
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
...
+            if (socket) {
+                for (int i = 0; i < beerocks::IFACE_TOTAL_VAPS; ++i) {
+                    if (socket->vaps_list.vaps[i].mac != network_utils::ZERO_MAC) {
+                        bssid_list.push_back(socket->vaps_list.vaps[i].mac);
+                    }
+                }
+            }
+        }
+
+        send_slave_ap_metrcic_query_message(0, bssid_list);
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: There is a typo in the name of the method

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 4]
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
...
+        for (auto &socket : slaves_sockets) {
+            if (socket) {
+                for (int i = 0; i < beerocks::IFACE_TOTAL_VAPS; ++i) {
+                    if (socket->vaps_list.vaps[i].mac != network_utils::ZERO_MAC) {
+                        bssid_list.push_back(socket->vaps_list.vaps[i].mac);
+                    }
+                }
+            }
+        }
+
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Do not send message if bssid_list is empty.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -2279,6 +2303,13 @@ bool backhaul_manager::handle_multi_ap_policy_config_request(ieee1905_1::CmduMes
         ap_metrics_reporting_info.reporting_interval_s =
             metric_reporting_policy_tlv->metrics_reporting_interval_sec();
         ap_metrics_reporting_info.last_reporting_time_point = std::chrono::steady_clock::now();
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I think this is the equivalent to your static ap_metric_interval. If true, we should use one or the other, but not both. IMO, we should use member variables and not static variables.

**Review by mariomaz — CHANGES_REQUESTED** [review]
I think there's some duplicate code.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 4 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 3 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Opened new issue https://github.com/prplfoundation/prplMesh/issues/1341

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 1 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed

**Comment by VladyslavTupikin** [comment]
This PR should be merged after https://github.com/prplfoundation/prplMesh/pull/1285

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: It doesn't make sense to open an issue for splitting things in a separate PR or changing commits - once the PR is merged, it becomes irrelevant.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 6]
```diff
@@ -2302,6 +2317,13 @@ bool backhaul_manager::handle_multi_ap_policy_config_request(ieee1905_1::CmduMes
...
+
+        if (ap_metrics_reporting_info.reporting_interval_s != 0) {
+            set_select_timeout(ap_metrics_reporting_info.reporting_interval_s *
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I think that changing the select timeout will have side effects on the rest of actions that have to be executed periodically.

**Review by mariomaz — APPROVED** [review]
I think that changing the select timeout when reporting interval is set is affecting other periodic actions.
Nevertheless, since it is a very small change in case you agree with me, I'm approving.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 6 above)
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Good point! The select timeout is currently set at a fixed 500ms. So there actually is no reason to change it since the reporting interval is going to be more than 1s. So @VladyslavTupikin please remove the set_select_timeout and add a comment that it's not needed since the timeout is 500ms.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 7]
```diff
@@ -1208,6 +1208,54 @@ bool backhaul_manager::send_autoconfig_search_message(std::shared_ptr<sRadioInfo
     LOG(DEBUG) << "sending autoconfig search message, bridge_mac=" << bridge_info.mac;
     return send_cmdu_to_bus(cmdu_tx, network_utils::MULTICAST_1905_MAC_ADDR, bridge_info.mac);
 }
+bool backhaul_manager::send_slave_ap_metric_query_message(uint16_t mid,
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: (nitpick) please leave an empty line before the function.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 8]
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
             if (elapsed_time_s >= ap_metrics_reporting_info.reporting_interval_s) {
                 ap_metrics_reporting_info.last_reporting_time_point = now;
 
-                // TODO: report AP metrics
+                std::vector<sMacAddr> bssid_list;
+                for (auto &socket : slaves_sockets) {
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
                for (const auto &socket : slaves_sockets) {
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 9]
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
...
+                    if (socket) {
+                        for (int i = 0; i < beerocks::IFACE_TOTAL_VAPS; ++i) {
+                            if (socket->vaps_list.vaps[i].mac != network_utils::ZERO_MAC) {
+                                bssid_list.push_back(socket->vaps_list.vaps[i].mac);
+                            }
+                        }
+                    }
+                }
+                if (!bssid_list.empty()) {
+                    send_slave_ap_metric_query_message(0, bssid_list);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1034]: I think this deserves a comment to explain the 0:
```suggestion
                    // We must generate a new MID for the response, so set MID to 0 here.
                    send_slave_ap_metric_query_message(0, bssid_list);
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 10]
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
...
+                } else {
+                    LOG(DEBUG)
+                        << "Failed to send AP_METRICS_QUERY to the son slave, empty BSSID list";
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: It's not really a failure...
```suggestion
                        << "Skipping AP_METRICS_QUERY for slave " << socket->hostap_iface <<", empty BSSID list";
```

**Review by arnout — APPROVED** [review]
Minor tweaks only.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 6 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed, added comment.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 7 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 8 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1034]: Added

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 10 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed just commentary, bssid_list contains all BSSID on all radio
